### DISCUSSION TITLE
Problems integrating with the Elm language server

### DISCUSSION THREAD

**Discussion by hovsater** [helix-editor/helix#10274] [Troubleshooting]
When writing Elm code, sometimes the LSP integration becomes erratic/broken and generates diagnostics for problems that don't exist. E.g., only partially interpreting identifiers, generating parsing errors and what not. Initially I thought it was due to incremental synchronziation, see https://github.com/elm-tooling/elm-language-server/issues/503, but I don't believe that's the case any more.

**Video of me making some edits and saving the file, triggering a formatting request.**
https://asciinema.org/a/dexV5nWbBn5sQ9iDRD5X2r9qB and here's the associated [log](https://gist.github.com/kevinsjoberg/ca0bb168748f91ec73d2f1ea0c703bc6).

**Vide of me making some edits and saving the file. Then manually triggering formatting.**
https://asciinema.org/a/Bh47XkiludVBgxK6U8b5LuJal and here's the associated [log](https://gist.github.com/kevinsjoberg/0e186bdf1af91e34fae71e9d87f26a37).

The videos above were recorded using the `master` build of Helix, Elm 0.19.1, elm-language-server 2.4.0 and Node v17.3.0.

I've [reported](https://github.com/elm-tooling/elm-language-server/issues/503#issuecomment-1019521420) the issue to [elm-language-server](https://github.com/elm-tooling/elm-language-server) as well.

**Comment by hovsater** [comment]
I just tried to replicate the videos above in [Sublime Text](https://www.sublimetext.com) but failed to do so. I do not get the parsing error diagnostic when performing the same operations in Sublime Text.

**Comment by archseer** [comment]
Does this only happen from formatting? Or can you replicate only by typing too?

**Comment by hovsater** [comment]
I have not been able to reproduce it when formatting isn't taking place.

**Comment by archseer** [comment]
There's also #1364, maybe we need to double check our formatting implementation.

**Comment by archseer** [comment]
Could it be that the Elm LSP isn't expecting `text/didChange` from formatting changes?

In your logs, this is the changes we get from the formatting request
```
2022-01-23T17:59:48.184 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","id":1,"result":[{"newText":"","range":{"start":{"line":47,"character":0},"end":{"line":47,"character":8}}},{"newText":"\n           ","range":{"start":{"line":48,"character":16},"end":{"line":48,"character":16}}}]}
```

Then we apply them and report them back

```
2022-01-23T17:59:48.198 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"textDocument/didChange","params":{"contentChanges":[{"range":{"end":{"character":8,"line":47},"start":{"character":0,"line":47}},"text":""},{"range":{"end":{"character":16,"line":48},"start":{"character":16,"line":48}},"text":"\n           "}],"textDocument":{"uri":"file:///Users/kevinsjoberg/code/shape_matcher/src/Main.elm","version":22}}}
```

The changes seem identical to me to what we got from the LSP so I don't see how helix could be wrong.
